spring-cloud-skipper/src/main/java/org/springframework/cloud/skipper/domain/deployer/ApplicationManifestDifference.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.cloud.skipper.domain.deployer;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.MapDifference;
+
+import org.springframework.util.Assert;
+
+/**
+ * Describes the difference between two ApplicationManifests
+ * @author Mark Pollack
+ */
+public class ApplicationManifestDifference {
+
+	private final String applicationName;
+
+	private final MapDifference apiAndKindDifference;
+
+	private final MapDifference metadataDifference;
+
+	private final MapDifference resourceAndVersionDifference;
+
+	private final MapDifference applicationPropertiesDifference;
+
+	private final MapDifference deploymentPropertiesDifference;
+
+	@JsonCreator
+	public ApplicationManifestDifference(@JsonProperty("applicationName") String applicationName,
+			@JsonProperty("apiAndKindDifference") MapDifference apiAndKindDifference,
+			@JsonProperty("metadataDifference") MapDifference metadataDifference,
+			@JsonProperty("resourceAndVersionDifference") MapDifference resourceAndVersionDifference,
+			@JsonProperty("applicationPropertiesDifference") MapDifference applicationPropertiesDifference,
+			@JsonProperty("deploymentPropertiesDifference") MapDifference deploymentPropertiesDifference) {
+		Assert.notNull(applicationName, "applicationName can not be null");
+		Assert.notNull(apiAndKindDifference, "apiAndKindDifference can not be null");
+		Assert.notNull(metadataDifference, "metadataDifference can not be null");
+		Assert.notNull(resourceAndVersionDifference, "resourceAndVersionDifference can not be null");
+		Assert.notNull(applicationPropertiesDifference, "applicationPropertiesDifference can not be null");
+		Assert.notNull(deploymentPropertiesDifference, "deploymentPropertiesDifference can not be null");
+		this.applicationName = applicationName;
+		this.apiAndKindDifference = apiAndKindDifference;
+		this.metadataDifference = metadataDifference;
+		this.resourceAndVersionDifference = resourceAndVersionDifference;
+		this.applicationPropertiesDifference = applicationPropertiesDifference;
+		this.deploymentPropertiesDifference = deploymentPropertiesDifference;
+	}
+
+	public String getApplicationName() {
+		return this.applicationName;
+	}
+
+	public MapDifference getApiAndKindDifference() {
```

Comment:
If this is somehow exposed via rest as json, what's the actual format? I know that next step would be to remove guava but I'm slightly worried what we expect as json.

Reply:
I remember looking at the fields in MapDifference once and thinking it was going to marshall/unmarshall to json well....  let's have a look.

Reply:
It has Maps as private fields and a ValueDifference object, which is a simple pair of primitives.  I haven't tested but we should be able to make an equivalent version of this that ser/deser's correctly over JSON.